File: Popup/Dtos/PopupResponseDto.cs
using System.Text.Json;
using System;

namespace Popup.Dtos
{
    /*
     * 서버에서 내려오는 팝업 조회 응답을 담는 DTO
     *
     * DTO는 JSON 데이터를 전달받기 위한 객체다.
     * 화면을 직접 그리거나 팝업을 여는 역할은 하지 않는다.
     */
    public class PopupResponseDto
    {
         /*
         * 팝업 노출 시작 일시
         *
         * null이면 시작 일시 제한 없이
         * 바로 노출할 수 있다.
         */
        public DateTimeOffset? DisplayStartAt { get; set; }

        /*
         * 팝업 노출 종료 일시
         *
         * null이면 종료 일시 제한 없이
         * 계속 노출할 수 있다.
         */
        public DateTimeOffset? DisplayEndAt { get; set; }


        /*
         * 팝업 고유 번호
         *
         * 나중에 조회, 로그 저장,
         * 다시 보지 않기 처리 등에 사용한다.
         */
        public long PopupId { get; set; }

        /*
         * 팝업 종류
         *
         * 서버 JSON 예:
         * TEXT
         * IMAGE
         * VIDEO
         * SURVEY
         * QUIZ
         */
        public string PopupType { get; set; } =
            string.Empty;

        /*
         * PopupWindow 상단에 표시할 제목
         */
        public string Title { get; set; } =
            string.Empty;

        /*
         * 팝업 크기 계산 방식
         *
         * 서버 JSON 예:
         * FIXED
         * VIEWPORT_RATIO
         * AUTO
         */
        public string SizeMode { get; set; } =
            "FIXED";

        /*
         * Fixed 모드에서 사용하는 고정 크기
         */
        public double Width { get; set; } =
            900;

        public double Height { get; set; } =
            620;

        /*
         * ViewportRatio 모드에서 사용하는
         * 모니터 작업 영역 대비 크기 비율
         */
        public double WidthRatio { get; set; } =
            0.7;

        public double HeightRatio { get; set; } =
            0.75;

        /*
         * 동적 계산 결과에 적용할 최소 크기
         */
        public double MinimumWidth { get; set; } =
            480;

        public double MinimumHeight { get; set; } =
            320;

        /*
         * 동적 계산 결과에 적용할 최대 크기
         */
        public double MaximumWidth { get; set; } =
            1200;

        public double MaximumHeight { get; set; } =
            900;

        /*
         * PopupWindow 공통 영역 표시 옵션
         */
        public bool ShowHeader { get; set; } =
            true;

        public bool ShowCloseButton { get; set; } =
            true;

        public bool ShowFooter { get; set; } =
            true;

        public bool ShowDoNotShowAgain { get; set; }

        /*
         * 팝업 종류별 상세 데이터
         *
         * TEXT, IMAGE, VIDEO, SURVEY마다
         * content 내부 구조가 다르므로
         * 공통 DTO에서는 JsonElement로 받아둔다.
         *
         * 이후 PopupFactory에서 PopupType을 확인하고
         * 각각의 ContentDto로 다시 변환한다.
         */
        public JsonElement Content { get; set; }
    }
}

File: popup-api/src/main/java/com/oksy/popup/dto/PopupResponseDto.java
package com.oksy.popup.dto;

import java.time.OffsetDateTime;
import java.util.Map;

/*
 * WPF 프로그램에 전달할
 * 팝업 한 건의 응답 정보를 담는 DTO다.
 *
 * 각 필드 이름은 WPF의 PopupResponseDto와
 * 동일한 camelCase JSON으로 변환된다.
 */
public record PopupResponseDto(

        /*
         * 팝업을 구분하는 고유 ID다.
         *
         * 30일간 보지 않기,
         * 팝업 조회 및 로그 저장에 사용한다.
         */
        String popupId,

        /*
         * 팝업 종류다.
         *
         * TEXT
         * IMAGE
         * VIDEO
         * SURVEY
         * QUIZ
         */
        String popupType,

        /*
         * PopupWindow 상단에 표시할 제목이다.
         */
        String title,

        /*
         * 팝업 노출 시작 일시다.
         *
         * null이면 시작 일시 제한 없이 표시할 수 있다.
         */
        OffsetDateTime displayStartAt,

        /*
         * 팝업 노출 종료 일시다.
         *
         * null이면 종료 일시 제한 없이 표시할 수 있다.
         */
        OffsetDateTime displayEndAt,

        /*
         * 여러 팝업을 표시하는 방식이다.
         *
         * SEQUENTIAL
         * → 순차 표시
         *
         * SIMULTANEOUS
         * → 동시 표시
         */
        String displayMode,

        /*
         * 팝업 크기를 계산하는 방식이다.
         *
         * FIXED
         * VIEWPORT_RATIO
         * AUTO
         */
        String sizeMode,

        /*
         * FIXED 방식에서 사용하는 고정 너비다.
         */
        double width,

        /*
         * FIXED 방식에서 사용하는 고정 높이다.
         */
        double height,

        /*
         * 작업 영역 대비 팝업 너비 비율이다.
         */
        double widthRatio,

        /*
         * 작업 영역 대비 팝업 높이 비율이다.
         */
        double heightRatio,

        /*
         * 팝업이 작아질 수 있는 최소 너비다.
         */
        double minimumWidth,

        /*
         * 팝업이 작아질 수 있는 최소 높이다.
         */
        double minimumHeight,

        /*
         * 팝업이 커질 수 있는 최대 너비다.
         */
        double maximumWidth,

        /*
         * 팝업이 커질 수 있는 최대 높이다.
         */
        double maximumHeight,

        /*
         * PopupWindow 상단 Header 표시 여부다.
         */
        boolean showHeader,

        /*
         * 상단 닫기 버튼 표시 여부다.
         */
        boolean showCloseButton,

        /*
         * PopupWindow 하단 Footer 표시 여부다.
         */
        boolean showFooter,

        /*
         * 30일간 보지 않기 영역 표시 여부다.
         */
        boolean showDoNotShowAgain,

        /*
         * 팝업 종류별 상세 데이터다.
         *
         * 팝업 종류마다 내부 구조가 다르므로
         * Map으로 받아서 JSON 객체 형태로 전달한다.
         *
         * TEXT
         * → contentTitle, description 등
         *
         * IMAGE
         * → imageUrl, imageTitle 등
         *
         * VIDEO
         * → videoUrl, videoTitle 등
         */
        Map<String, Object> content
) {
}